### ISSUE TITLE
hessianstructure / second order derivative test inconsistency

### ISSUE THREAD

**Issue by thartland** [mechmotum/cyipopt#143]
I am using cyipopt for a project and attempted to use a _dense Hessian structure_ but the finite difference check indicates that the objective Hessian is inconsistent with the finite-difference check for the gradient. When I use a _lower-triangular Hessian structure_ for the Hessian as in [hs071.py](https://github.com/mechmotum/cyipopt/blob/master/examples/hs071.py) the finite difference check does not find any inconsistencies with the objective Hessian and gradient.

I had assumed that using either a dense or lower triangular structure would be ok, after browsing the [Ipopt Triplet Format for Sparse Matrices](https://coin-or.github.io/Ipopt/IMPL.html#TRIPLET). To avoid future errors I would appreciate if someone could help me clarify my misunderstanding.

The code which reproduces the results I described above can be found in the attached text file 

[simple_example.txt](https://github.com/mechmotum/cyipopt/files/7681943/simple_example.txt)

**Comment by thartland** [comment]
I should have mentioned that when I use the dense Hessian structure, the off-diagonal elements of the Hessian are twice as large as that predicted by the finite-difference check.

**Comment by moorepants** [comment]
Here's the example code for easier copy-paste:

```python
import numpy as np
import cyipopt


class example:

    def __init__(self, A, trilstruct=True):
        self.A = A
        self.trilstruct = trilstruct

    def objective(self, x):
        return 0.5*np.inner(x, self.A.dot(x))

    def gradient(self, x):
        return self.A.dot(x)

    def constraints(self, x):
        return x[0] - 1.

    def jacobian(self, x):
        return np.array((1., 0., 0., 0.))

    def hessianstructure(self):
        if self.trilstruct:
            return np.nonzero(np.tril(np.ones((4, 4))))
        else:
            return np.nonzero(np.ones((4,4)))

    def hessian(self, x, lagrange, obj_factor):
        #
        # The callback for calculating the Hessian
        #
        H = obj_factor*self.A.copy()

        row, col = self.hessianstructure()
        return H[row, col]

    def intermediate(
            self,
            alg_mod,
            iter_count,
            obj_value,
            inf_pr,
            inf_du,
            mu,
            d_norm,
            regularization_size,
            alpha_du,
            alpha_pr,
            ls_trials
            ):

        #
        # Example for the use of the intermediate callback.
        #
        print("Objective value at iteration #%d is - %g" % (iter_count, obj_value))


def main():
    #
    # Define the problem
    #

    x0 = [1.0, 5.0, 5.0, 1.0]

    lb = [0.0, 0.0, 0.0, 0.0]
    ub = [7.0, 7.0, 7.0, 7.0]

    cl = [0.]
    cu = [0.]
    a = 0.01
    A = np.array(((1.0, a, 0.0, 0.0),
                  (a, 1.0, 0.0, 0.0),
                  (0.0, 0.0, 1.0, 0.0),
                  (0.0, 0.0, 0.0, 1.0)))


    for trilstruct in [False, True]:
        print("----- using lower-triangular Hessian structure? "+str(trilstruct)+"------")
        nlp = cyipopt.Problem(
            n=len(x0),
            m=len(cl),
            problem_obj=example(A, trilstruct=trilstruct),
            lb=lb,
            ub=ub,
            cl=cl,
            cu=cu
            )

        #
        # Set solver options
        #
        nlp.add_option('derivative_test', 'only-second-order')
        nlp.add_option('max_iter', 0)
        #
        # Solve the problem
        #
        x, info = nlp.solve(x0)



if __name__ == '__main__':
    main()
```

**Comment by thartland** [comment]
Here is the specific output for the second-derivative tests for further completeness.

1. When a dense Hessian structure is used:

Starting derivative checker for second derivatives with obj_factor or lambda[i] set to 1.5.
obj_hess[    0,    1] =  2.9999999999999999e-02 v  ~  1.5000001241105565e-02  [ 1.000e+00]
obj_hess[    1,    0] =  2.9999999999999999e-02 v  ~  1.4999998395035838e-02  [ 1.000e+00]
Derivative checker detected 2 error(s).

2. When a lower-triangular Hessian structure is used:

Starting derivative checker for second derivatives with obj_factor or lambda[i] set to 1.5.
No errors detected by derivative checker.

I am now assuming that since the Hessian is symmetric the off-diagonal elements are "duplicated", which would be consistent with having off-diagonal objective Hessian elements that are twice as large as they should be. Is this correct?

**Comment by thartland** [comment]
All issues of mine have been resolve, now assuming that the off-diagonal elements are duplicated and so that a user should only supply the sparsity structure for either the upper or lower triangle of the objective/constraint Hessians. This was not an appropriate place to post this question.

**Comment by moorepants** [comment]
Are you saying this was an invalid input?

**Comment by thartland** [comment]
What I understand now is that if the user provides a Hessian element H_(i,j) i < j, then ipopt knows to duplicate this element H_(j,i) = H_(i,j) internally. And so when one provides both H_(i,j) and H_(j,i) as I did via specifying the Hessian structure as np.nonzeros(np.ones((4,4))), then the off-diagonal elements are both duplicated and ipopt sees these elements twice as large as they should be. This result is consistent with the derivative error check results I posted above  


    1. When a dense Hessian structure (np.nonzero(np.ones((4,4)))) is used:

Starting derivative checker for second derivatives with obj_factor or lambda[i] set to 1.5.
obj_hess[ 0, 1] = 2.9999999999999999e-02 v ~ 1.5000001241105565e-02 [ 1.000e+00]
obj_hess[ 1, 0] = 2.9999999999999999e-02 v ~ 1.4999998395035838e-02 [ 1.000e+00]
Derivative checker detected 2 error(s).

The error was on my end, in that I did not know that ipopt repeated off-diagonal elements of the objective/constraint Hessians and so one only needs to supply the nonzero elements that are not above the main diagonal.
